crates/core/src/lib.rs

```diff
@@ -16,17 +16,605 @@ pub mod network;
+
+        self.with_networking(networking, metadata_builder)
+    }
+}
+
+pub struct CoreBuilderWithNetworking<T> {
+    metadata_store_client: MetadataStoreClient,
+    node_name: String,
+    force_node_id: Option<NodeId>,
+    advertise_address: AdvertisedAddress,
+    roles: EnumSet<Role>,
+    nodes_config: NodesConfiguration,
+    metadata_manager: MetadataManager<T>,
+    metadata: Metadata,
+    networking: Networking<T>,
+}
+
+impl<T> CoreBuilderWithNetworking<T> {
+    pub fn with_tc(self, tc: TaskCenter) -> CoreBuilderFinal<T> {
+        CoreBuilderFinal {
+            tc,
+            node_name: self.node_name,
+            force_node_id: self.force_node_id,
+            advertise_address: self.advertise_address,
+            roles: self.roles,
+            nodes_config: self.nodes_config,
+            metadata_manager: self.metadata_manager,
+            metadata_store_client: self.metadata_store_client,
+            networking: self.networking,
+            metadata: self.metadata,
+            provider_kind: ProviderKind::Local,
+            allow_bootstrap: false,
+            partition_table: PartitionTable::with_equally_sized_partitions(Version::MIN, 10),
+            replication_strategy: ReplicationStrategy::OnAllNodes,
+            router_builder: MessageRouterBuilder::default(),
+        }
+    }
+}
+
+pub struct CoreBuilderFinal<T> {
+    pub tc: TaskCenter,
+    node_name: String,
+    force_node_id: Option<NodeId>,
+    advertise_address: AdvertisedAddress,
+    roles: EnumSet<Role>,
+    pub metadata_manager: MetadataManager<T>,
+    pub metadata_store_client: MetadataStoreClient,
+    pub metadata: Metadata,
+    pub router_builder: MessageRouterBuilder,
+    partition_table: PartitionTable,
+    replication_strategy: ReplicationStrategy,
+    pub nodes_config: NodesConfiguration,
+    provider_kind: ProviderKind,
+    pub networking: Networking<T>,
+    allow_bootstrap: bool,
+}
+
+impl<T: TransportConnect> CoreBuilderFinal<T> {
+    pub fn set_num_partitions(mut self, num_partitions: u16) -> Self {
+        self.partition_table = PartitionTable::with_equally_sized_partitions(
+            self.partition_table.version(),
+            num_partitions,
+        );
+        self
+    }
+
+    pub fn set_replication_strategy(mut self, replication_strategy: ReplicationStrategy) -> Self {
+        self.replication_strategy = replication_strategy;
+        self
+    }
+
+    pub fn set_provider_kind(mut self, provider_kind: ProviderKind) -> Self {
+        self.provider_kind = provider_kind;
+        self
+    }
+
+    pub fn set_allow_bootstrap(mut self, allow_bootstrap: bool) -> Self {
+        self.allow_bootstrap = allow_bootstrap;
+        self
+    }
+
+    pub fn set_nodes_config(
+        mut self,
+        node_name: impl Into<String>,
+        node_id: GenerationalNodeId,
+        nodes_config: NodesConfiguration,
+    ) -> Self {
```

Comment:
I found this confusing when reading the code at call-sites. It wasn't clear what the other two arguments mean. I realised eventually that it's to set my node name and id. We can keep this as is and remove the `node_id`, since we should be able to find the ID from the config. Or split those into their own setters maybe?